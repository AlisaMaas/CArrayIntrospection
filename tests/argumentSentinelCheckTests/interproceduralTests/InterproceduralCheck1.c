/**
 * This is a simple interproceedural test that ensures with three functions,
 * it recognizes that foo calls the null-terminated print though it contains no
 * null checks. The results should note that string is null terminated in both foo
 * and find, but not print.
 **/
int find(char[]);
void print(char *string) {}
int foo(char string[]) {
	for (int i = 0;; i++) {
		if (string[i] != '\0') {
			print("About to break from loop");
			break;
		}
	}
	return find(string);
}
int find(char string[]) {
	for (int i = 0;; i++) {
		if (string[i] == '\0') {
			print("About to break from loop");
			break;
		}
	}
	return 1;
}

int bar(char* a, int n) {
  for (int i = 0; i < n; i++) {
    print(&a[i]);
  }
  return 1;
}