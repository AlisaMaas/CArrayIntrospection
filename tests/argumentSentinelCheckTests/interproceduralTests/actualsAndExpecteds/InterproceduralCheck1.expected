Printing analysis 'Promote Memory to Register' for function 'print':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Promote Memory to Register' for function 'foo':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Promote Memory to Register' for function 'find':
Pass::print not implemented for pass: 'Promote Memory to Register'!
Printing analysis 'Determine whether and how to annotate each function with the null-terminated annotation':
foo with argument 0 should be annotated NULL_TERMINATED (2)  because  found a call to find passing string
find with argument 0 should be annotated NULL_TERMINATED (2)  because found a non optional sentinel check for string in find
